applied-crypto-examples/src/encrypted_zksnark.rs

```diff
@@ -0,0 +1,254 @@
+/// scalars by the challenge curve points
+pub struct EncryptedResponse {
+    /// Evaluation of all polynomial coefficients at the challenge curve points
+    pub px: RistrettoPoint,
+    /// Evaluation of h(s) = p(s)/t(s) at the challenge curve points
+    pub hx: RistrettoPoint,
+}
+
+impl EncryptedResponse {
+    /// New encrypted response object
+    pub fn new(px: RistrettoPoint, hx: RistrettoPoint) -> Self {
+        Self { px, hx }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Proposed roots would result in a polynomial with coefficients in the rational field
+    OutsideIntegerField(i64, i64),
+    /// No public roots set
+    NoPublicRoots,
+}
+
+/// Polynomial with coefficients restricted to integers within the field of 8-bit signed integers
+#[derive(Clone, Debug)]
+pub struct Polynomial {
+    // Polynomial roots (a, b) such that a*x + b is a factor of the polynomial
+    roots: Vec<Root>,
+    // Polynomial coefficients
+    coefficients: Vec<Scalar>,
+    // Number of public roots
+    num_public_roots: usize,
+}
+
+impl Polynomial {
+    /// Create a new polynomial from a list of roots
+    pub fn new(roots: Vec<Root>, num_public_roots: usize) -> Result<Self, Error> {
+        if num_public_roots == 0 {
+            return Err(Error::NoPublicRoots);
+        }
+        let mut coefficients = Vec::new();
+        for root in roots.iter() {
+            if coefficients.is_empty() {
+                coefficients.push(root.a);
+                coefficients.push(root.b);
+            } else {
+                let mut new_coefficients = Vec::new();
+                new_coefficients.push(coefficients[0] * root.a);
+                for i in 1..coefficients.len() {
+                    new_coefficients.push(coefficients[i - 1] * root.b + coefficients[i] * root.a);
+                }
+                new_coefficients.push(coefficients[coefficients.len() - 1] * root.b);
+                coefficients = new_coefficients;
+            }
+            //println!("coefficients: {:?}", coefficients);
+        }
+        coefficients.reverse();
+        Ok(Self {
+            roots,
+            coefficients,
+            num_public_roots,
+        })
+    }
+
+    /// Get degree of polynomial
+    pub fn degree(&self) -> usize {
+        self.roots.len()
+    }
+
+    /// calculate encrypted polynomial
+    pub fn generate_response(&self, powers: &EncryptedChallenge) -> EncryptedResponse {
+        let px = self.eval(powers);
+        let h = Polynomial::new(self.roots[self.num_public_roots..].to_vec(), 1).unwrap();
+        let hx = h.eval(powers);
+        EncryptedResponse::new(px, hx)
+    }
+
+    /// Evaluate polynomial at given encrypted powers
+    pub fn eval(&self, powers: &EncryptedChallenge) -> RistrettoPoint {
+        powers
+            .encrypted_powers
+            .iter()
+            .zip(self.coefficients.iter())
+            .map(|(p, c)| p * c)
+            .sum()
+    }
```

Comment:
This should probably be made non-public, and tests of correct polynomial evaluation should be made another way.